### ISSUE TITLE
Allow deletion of failed pods

### ISSUE THREAD

**Issue by jaymebrd** [kubernetes-sigs/descheduler#596]
**Is your feature request related to a problem? Please describe.**

When pods are in status `Failed` the only mechanism to clean up these pods is the kubernetes[ garbage collector](https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/#pod-garbage-collection) which is currently defaulted to 12500 which means that for most people this is useless.

There's been a few recent issues where the reason for the pod failures has been related to a kubernetes bug. One particular example is the recent nodeAffinity issue where the Kubelet rejects pod scheduled based on newly added node labels which have not been observed by the kubelet yet #93338

```
Status:               Failed
Reason:               NodeAffinity
Message:              Pod Predicate NodeAffinity failed
```

This can be particularly problematic to certain custom controllers which haven't implemented recreation of failed pod, or deletion of failed pods.

**Describe the solution you'd like**

Allow deletion of failed pods, possibly based on some reason exclusions (you may want to keep evicted pods around, for example)

**Describe alternatives you've considered**

The default garbage collection value in kubernetes could be made configurable, and/or given a better default than 12500, however, it seems more appropriate for a project like this to allow more configuration on which `reason` should be inclusive/exclusive of the automatic deletion.

**What version of descheduler are you using?**

descheduler version: N/A

**Additional context**

See more details here for a related the feature request made to kubernetes. https://github.com/kubernetes/kubernetes/issues/99986
This feels like a potential good fit for the descheduler.

**Comment by damemi** [comment]
This is a hardcoded filter the descheduler uses when listing pods on a node: https://github.com/kubernetes-sigs/descheduler/blob/84b174e8417edbb13e51a3207191c3449a4ce309/pkg/descheduler/pod/pods.go#L83

So it's not just that it won't evict them, but it doesn't even see them at all. This makes sense because if it listed Failed pods, it would consider them in each strategy when calculating pods to evict, which could produce odd behavior (for example, evicting a pod for antiaffinity with a failed pod on the same node doesn't really make sense)

However, I see some of the use cases you have pointed out as valid. So we could add some (non-default) option to evict, or at least consider, Failed pods, but we should think of the best way to do that.

What about if we add a new strategy specifically for this kind of garbage collection? Something like `RemoveFailedPods`. Simple enough and doesn't interfere with core code that would affect the other strategies which don't really need failed pods in their logic

**Comment by jaymebrd** [comment]
> What about if we add a new strategy specifically for this kind of garbage collection? Something like RemoveFailedPods. Simple enough and doesn't interfere with core code that would affect the other strategies which don't really need failed pods in their logic

This makes a lot of sense to me and is effectively the approach I had to take with a custom controller to overcome the `NodeAffinity` failed pod issue. In my use-case I'm happy to remove all pods in failed state, but I could imagine it might be nice to specify the `Reason` either as an allow list or deny list for some people so that not all failed pods are automatically deleted.

**Comment by damemi** [comment]
> I could imagine it might be nice to specify the Reason either as an allow list or deny list for some people so that not all failed pods are automatically deleted.

This could be added as a parameter later on, if there is demand for it

**Comment by jaymebrd** [comment]
I'd be happy to try implement something in the next few weeks if you'd like? If you're okay with that, any pointers would be great.

**Comment by a7i** [comment]
@jaymebrd @damemi  I took a stab at this in PR #610 
Look forward to your feedback!

As far as the `reasons`, `minPodLifeTimeSeconds`, `excludeOwnerKinds` params, we are planning on moving off an in-house remediator tool to Descheduler and need these filters.

>`minPodLifeTimeSeconds` - Leave Pods around for n seconds to allow debugging
`excludeOwnerKinds` - Exclude Jobs

**Comment by Levovar** [comment]
Will this policy also handle Pods which errored out due to Topology Manager not being involved in scheduling decisions? See Known Limitations chapter under https://kubernetes.io/docs/tasks/administer-cluster/topology-manager/
AFAIK "Topology Affinity Error" is a distinct status on its own (i.e. not a "reason" of a "PodFailed" status), but it would be great if this strategy could be used for these Pods too!

**Comment by a7i** [comment]
@Levovar Would you be able to share a YAML/JSON showing the status of the Pod? More than happy to implement this if you can provide that.

**Comment by Levovar** [comment]
@a7i I was just looking at related issues where the error appeared in the "Status" column such as in https://github.com/kubernetes/kubernetes/issues/84597#issue-515432349 and https://github.com/kubernetes/kubernetes/issues/84869#issue-518618378
code-wise it also seems this error is returned directly to Kubelet in the Admit interface call result: https://github.com/kubernetes/kubernetes/blob/617064d7327bf455955cb785d57330c425d48b51/pkg/kubelet/cm/topologymanager/scope_pod.go#L55

Note I'm not a K8s dev so this still might be just a "Reason" field being display, just wanted to make sure if this use-case is covered :) I will try and re-create this scenario when I can to get a full JSON output of the Pod's status fields

**Comment by Levovar** [comment]
maybe @swatisehgal could help us out
will "Topology Affinity Error"ed Pods picked up by the following filter:
`fieldSelectorString := status.phase=" + string(v1.PodFailed)`
aka are these Pods "Failed", or their phase will be "TopologyAffinityError"?

**Comment by seanmalloy** [comment]
I believe we can close this now that #610 is merged.

@Levovar please feel free to open a separate enhancement issue if the initial new strategy implementation does not work with your Topology Manager use case. Thanks!

This new strategy will be shipped with the descheduler v0.22.0 release.

/close
